optimum/intel/utils/modeling_utils.py

```diff
@@ -110,3 +111,85 @@ def _find_files_matching_pattern(
+
+    Example:
+    .. code-block:: python
+
+        from optimum.intel.ipex import IPEXModelForCausalLM
+        from optimum.intel.utils.modeling_utils import bind_cores_for_best_perf
+
+        bind_cores_for_best_perf()
+        model = IPEXModelForCausalLM.from_pretrained("gpt2", torch_dtype=torch.bfloat16, export=True)
+        tokenizer = AutoTokenizer.from_pretrained("gpt2")
+        input_sentence = ["tell me a story about a trip to the moon"]
+        model_inputs = tokenizer(input_sentence, return_tensors="pt")
+        generation_kwargs = dict(max_new_tokens=500)
+        generated_ids = model.generate(**model_inputs, **generation_kwargs)
+
+    Returns:
+        None
+
+    """
+
+    import importlib.util
+    import platform
+
+    system = platform.system()
+    if system == "Linux":
+        if importlib.util.find_spec("numa") is not None:
+            import math
+
+            import numa
+            import psutil
+
+            local_size = get_int_from_env(
+                ["MPI_LOCALNRANKS", "OMPI_COMM_WORLD_LOCAL_SIZE", "MV2_COMM_WORLD_LOCAL_SIZE"], 1
+            )
+            rank_id = get_int_from_env(
+                ["LOCAL_RANK", "MPI_LOCALRANKID", "OMPI_COMM_WORLD_LOCAL_RANK", "MV2_COMM_WORLD_LOCAL_RANK"], 0
+            )
+            nodes = numa.get_max_node() + 1
+            rank_per_node = math.ceil(local_size / nodes)
+            num_cpus_per_nodes = int(psutil.cpu_count(logical=False) / nodes)
+            node_id = int(rank_id / rank_per_node)
+            rank_offset_per_node = rank_id % rank_per_node
+            if os.getenv("OMP_NUM_THREADS") is None:
+                # set OMP_NUM_THREADS to num of physical cores per socket
+                num_cpus_per_rank = max(int(num_cpus_per_nodes / rank_per_node), 1)
+                print("setting OMP_NUM_THREADS to", num_cpus_per_rank)
+            else:
+                num_cpus_per_rank = int(os.getenv("OMP_NUM_THREADS"))
+                print("OMP_NUM_THREADS already set to ", num_cpus_per_rank)
+            if len(numa.get_membind()) == nodes:
+                # if numa memory binding is not set, set it to the node where the rank is running
+                numa.set_membind([node_id])
+
+            torch.set_num_threads(num_cpus_per_rank)
+
+            if len(numa.get_affinity(0)) == psutil.cpu_count(logical=True):
+                # if numa affinity is unset (default value is set to all logical cores) set it to the physical cores assigned to the rank
+                cpu_start = num_cpus_per_rank * rank_offset_per_node
+                numa.set_affinity(
+                    0,
+                    list(numa.node_to_cpus(node_id))[cpu_start : cpu_start + num_cpus_per_rank],
+                )
+            print(f"affinity={numa.get_affinity(0)}, membind = {numa.get_membind()}")
+        else:
+            print("numa module not found, skipping binding cores")
+    else:
+        print("OS not supported, skipping binding cores")
```

Comment:
raising an error here makes more sense for me

Reply:
I switched it to logger.error. I didn't raise an error to interrupt execution because even if the core binding won't get executed the code would still run.

Reply:
but that's too implicit, the function is supposed to do core binding, if a user calls it, they expect it to do that, if a package is missing or their system doesn't allow for it, the code should fail with a module not found error or os error.

Reply:
makes sense. I've raised Errors